### PR TITLE
Add Wasm testing in Chromium, not just command-line `v8`

### PR DESCRIPTION
**Pull request by directhex**
This should give us a better idea of any cases where browser and JS engine behaviour diverge.

### CODE REVIEW

**Comment by directhex** [comment]
Failures are because https://dnceng.visualstudio.com/internal/_git/dotnet-helix-machines/pullrequest/9677 hasn't landed in prod Helix instance yet.

File: eng/pipelines/runtime.yml
Comment (premun) [inline eng/pipelines/runtime.yml:339]: https://github.com/dotnet/arcade/issues/5919 should be fixed now, but you might need to update the Helix SDK since Arcade wasn't released yet

Comment (directhex) [reply eng/pipelines/runtime.yml:339]: I tried bumping Arcade as part of this for testing and EVERYTHING EXPLODED so I'll wait for a bump to hit master I think

Comment (premun) [reply eng/pipelines/runtime.yml:339]: @directhex bump Helix SDK, not arcade. Check the iOS draft PR and cherry pick that

File: eng/testing/tests.mobile.targets
Comment (akoeplinger) [inline eng/testing/tests.mobile.targets:17]: nit: I'd prefer substituting the whole command rather than appending only the `-browser` portion.

E.g.
```suggestion
    <RunScriptCommand>$HARNESS_RUNNER wasm $XHARNESS_COMMAND --app=. --engine=$(JSEngine) $(JSEngineArgs) --js-file=runtime.js -v --output-directory=$XHARNESS_OUT -- --run WasmTestRunner.dll $(AssemblyName).dll</RunScriptCommand>
```

and then setting `XHARNESS_COMMAND` to `test` or `test-browser` depending on Scenario.

**Comment by steveisok** [comment]
Definitely looks like we're still blocked by https://github.com/dotnet/runtime/issues/41160 and https://github.com/dotnet/runtime/issues/41269.

**Comment by radical** [comment]
Corresponding `xharness` PR - https://github.com/dotnet/xharness/pull/309 .

**Comment by steveisok** [comment]
@directhex Please bump xharness and try again when the xharness PR lands.

**Comment by directhex** [comment]
@steveisok yep, got my eye on it

File: src/mono/wasm/runtime-test.js
Comment (radical) [inline src/mono/wasm/runtime-test.js:21]: Um.. should this be:
```
if (typeof print === undefined)
   print = console.log;
```

.. for the non-browser case?

**Comment by radical** [comment]
`System.Xml.XmlSchemaSet`:

```
[18:25:38] info: test-browser[0]
      Tests run: 585, Errors: 0, Failures: 0, Skipped: 0. Time: 10.16845s
      
[18:25:38] dbug: test-browser[0]
      Failed trying to read log messages via selenium: System.ArgumentException: http://127.0.0.1:38247/index.html?arg=--run&arg=WasmTestRunner.dll&arg=System.Xml.XmlSchemaSet.Tests.dll&arg=-notrait&arg=category%3dIgnoreForCI&arg=-notrait&arg=category%3dOuterLoop&arg=-notrait&arg=category%3dfailing - Access to fetch at 'http://www.w3.org/2001/03/xml.xsd' from origin 'http://127.0.0.1:38247' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.
         at Microsoft.DotNet.XHarness.CLI.Commands.Wasm.WasmBrowserTestRunner.RunSeleniumLogMessagePump(IWebDriver driver, CancellationToken token) in /_/src/Microsoft.DotNet.XHarness.CLI/Commands/WASM/Browser/WasmBrowserTestRunner.cs:line 222

< /snip>

XHarness exit code:
XHarness artifacts: ./xharness-output
```

Exit code is timeout.

**Comment by directhex** [comment]
Wow, what an absolutely *phenomenal* improvement on the last run. Bravo, @radical

**Comment by directhex** [comment]
```
[18:25:02] dbug: test-browser[0]
      [FAIL] System.Xml.Tests.AsyncReaderLateInitTests.ReadAfterInitializationWithUriOnAsyncReaderTrows
      
[18:25:02] dbug: test-browser[0]
      Assert.Throws() Failure
      
[18:25:02] dbug: test-browser[0]
      Expected: typeof(System.Net.Http.HttpRequestException)
      
[18:25:02] dbug: test-browser[0]
      Actual:   typeof(System.PlatformNotSupportedException): Cannot wait on monitors on this runtime.
      
[18:25:02] dbug: test-browser[0]
      ---- System.PlatformNotSupportedException : Cannot wait on monitors on this runtime.
```

**Comment by radical** [comment]
`System.Xml`:

```
[18:25:02] dbug: test-browser[0]
      [FAIL] System.Xml.Tests.AsyncReaderLateInitTests.ReadAsyncAfterInitializationWithUriThrows
      System.InvalidOperationException : An asynchronous operation is already in progress.     
         at System.Xml.XmlAsyncCheckReader.CheckAsync()
         at System.Xml.XmlAsyncCheckReader.Dispose(Boolean disposing)
         at System.Xml.XmlReader.Dispose()
         at System.Xml.Tests.AsyncReaderLateInitTests.ReadAsyncAfterInitializationWithUriThrows()
         at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
```

```
[18:25:02] dbug: test-browser[0]
      [FAIL] System.Xml.Tests.AsyncReaderLateInitTests.ReadAfterInitializationWithUriOnAsyncReaderTrows

      Assert.Throws() Failure
      Expected: typeof(System.Net.Http.HttpRequestException)
      Actual:   typeof(System.PlatformNotSupportedException): Cannot wait on monitors on this runtime.
     
      ---- System.PlatformNotSupportedException : Cannot wait on monitors on this runtime.
         at System.Threading.Monitor.ObjWait(Int32 millisecondsTimeout, Object obj)
         at System.Threading.Monitor.Wait(Object obj, Int32 millisecondsTimeout)
         at System.Threading.ManualResetEventSlim.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
         at System.Threading.Tasks.Task.SpinThenBlockingWait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
         at System.Threading.Tasks.Task.InternalWaitCore(Int32 millisecondsTimeout, CancellationToken cancellationToken)
         at System.Threading.Tasks.Task.InternalWait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
         at System.Xml.XmlTextReaderImpl.FinishInitUriString()
         at System.Xml.XmlTextReaderImpl.FinishInit()
         at System.Xml.XmlTextReaderImpl.Read()
         at System.Xml.XmlAsyncCheckReader.Read()
         at System.Xml.Tests.AsyncReaderLateInitTests.<>c__DisplayClass8_0.<ReadAfterInitializationWithUriOnAsyncReaderTrows>b__0()

      ----- Inner Stack Trace -----
         at System.Threading.Monitor.ObjWait(Int32 millisecondsTimeout, Object obj)
         at System.Threading.Monitor.Wait(Object obj, Int32 millisecondsTimeout)
         at System.Threading.ManualResetEventSlim.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
         at System.Threading.Tasks.Task.SpinThenBlockingWait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
         at System.Threading.Tasks.Task.InternalWaitCore(Int32 millisecondsTimeout, CancellationToken cancellationToken)
         at System.Threading.Tasks.Task.InternalWait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
         at System.Xml.XmlTextReaderImpl.FinishInitUriString()
         at System.Xml.XmlTextReaderImpl.FinishInit()
         at System.Xml.XmlTextReaderImpl.Read()
         at System.Xml.XmlAsyncCheckReader.Read()
         at System.Xml.Tests.AsyncReaderLateInitTests.<>c__DisplayClass8_0.<ReadAfterInitializationWithUriOnAsyncReaderTrows>b__0()
      ```

**Comment by radical** [comment]
I'm working on the failures in `XmlSchemaTest`

**Comment by radical** [comment]
fix for the failures seen in `XmlSchemaSet` - https://github.com/dotnet/xharness/pull/311 , which were failing the step even though all the tests had passed.

**Comment by radical** [comment]
Should I open a separate PR for fixing `System.Net.Http/tests/FunctionalTests`? Or add it here?  https://gist.github.com/radical/f7f1a99ac9f4bd33554056401fdfeb4f

**Comment by steveisok** [comment]
I'd probably just add it here.

**Comment by radical** [comment]
`System.Net.Http`'s `FunctionalTests` can now be enabled by changing `SkipOnMono` in `src/libraries/System.Net.Http/tests/FunctionalTests/AssemblyInfo.cs`.

And https://github.com/dotnet/xharness/pull/313 - has some nice additions.

File: src/mono/wasm/runtime-test.js
Comment (lewing) [inline src/mono/wasm/runtime-test.js:13]: I want this in asap but after it lands I think we should consider sending all this as json over the websocket and expand it on the xharness side, we should be able to clean up the console output considerably.

pseudo code like
```
JSON.stringify ({ 'method': method, args: {...args}}) 
```
where method is { 'print', 'console.debug', ... }

and after that it might make sense to add a custom log provider so that send even more structure through.

File: src/mono/wasm/runtime-test.js
Comment (lewing) [inline src/mono/wasm/runtime-test.js:26]: This might be a better place to force everything through print?  I sometimes use runtime-tests.js outside of xharness

Comment (radical) [reply src/mono/wasm/runtime-test.js:26]: Yeah, that makes sense. And I was thinking maybe the WebSocket thing should be optional? I found it useful to run the tests without the harness, directly in a browser.

Comment (lewing) [reply src/mono/wasm/runtime-test.js:26]: I think it is worth a follow-up once we have everything working in CI.

**Comment by radical** [comment]
1 failing test remaining:

```
      [FAIL] System.Xml.Tests.AsyncReaderLateInitTests.ReadAsyncAfterInitializationWithUriThrows

      System.InvalidOperationException : An asynchronous operation is already in progress.     
         at System.Xml.XmlAsyncCheckReader.CheckAsync()
         at System.Xml.XmlAsyncCheckReader.Dispose(Boolean disposing)
         at System.Xml.XmlReader.Dispose()
         at System.Xml.Tests.AsyncReaderLateInitTests.ReadAsyncAfterInitializationWithUriThrows()
         at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
```

**Comment by steveisok** [comment]
@radical I'm ok with skipping the last failing test.  

@akoeplinger Can you give this a final review to move things forward?

**Review by steveisok — APPROVED** [review]

**Comment by lewing** [comment]
@steveisok can we review System.Xml.Tests.AsyncReaderLateInitTests in the for the actual browser case after this.

**Review by akoeplinger — APPROVED** [review]

File: src/libraries/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs
Comment (akoeplinger) [inline src/libraries/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs:946]: do we have a tracking issue for this?

Comment (lewing) [reply src/libraries/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs:946]: Should any of the SocketsHttpHandler tests be working?

Comment (marek-safar) [reply src/libraries/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs:946]: No, SocketsHttpHandler should throw PNSE on wasm-browser

**Comment by lewing** [comment]
Opened https://github.com/dotnet/runtime/issues/42852 to discuss the loopback server problem

File: src/libraries/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs
Comment (akoeplinger) [reply src/libraries/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs:946]: Ok, let's follow up on this.

**Pull request merged**
